File: src/bezier-curve-visualizer/src/dncNPoints.js
import { midP } from "./util";

/* Fungsi untuk mengiterasi list koordinat titik (points) hingga mendapatkan
 titik tengah hasil iterasi */
function iterateFind(points) {
    let lengthArr = points.length;
    let arr = points.slice(); 
    let leftArr = [points[0]];
    let rightArr = [points[lengthArr - 1]];
    let midPoints = [];

    while (lengthArr >= 2) {
        let arrTemp = [];
        for (let i = 0; i < lengthArr - 1; i++) {
            const midPoint = midP(arr[i], arr[i + 1]);
            arrTemp.push(midPoint);
        }
        leftArr.push(arrTemp[0]);
        rightArr.unshift(arrTemp[lengthArr - 2]);
        arr = arrTemp;
        lengthArr--;
        midPoints.push(arrTemp);
    }

    return [leftArr, [arr[0]], rightArr, midPoints];
}

/* Fungsi rekursif sebagai bentuk implementasi Divide and Conquer */
export function dncBezier(points, iteration, depth) {
    if (depth === 0) {
        return { points: [], midPointsHistory: [] };
    }

    const [leftSide, mid, rightSide, currentMidPoints] = iterateFind(points);
    let leftResult = dncBezier(leftSide, iteration, depth - 1);
    let rightResult = dncBezier(rightSide, iteration, depth - 1);

    return {
        points : (leftResult.points).concat(mid).concat(rightResult.points),
        midPointsHistory: (leftResult.midPointsHistory).concat([{ iteration: iteration-depth+1, midPoints: currentMidPoints }], rightResult.midPointsHistory)
    };
}

File: src/bezier-curve-visualizer/src/backup/dncNPoints.py
from util import Point, midP

def iterate_find(points):
    length_arr = len(points)
    arr = points[:]
    left_arr = [points[0]]
    right_arr = [points[-1]]
    mid_points = []

    while length_arr >= 2:
        arr_temp = []
        for i in range(length_arr - 1):
            mid_point = midP(arr[i], arr[i + 1])
            arr_temp.append(mid_point)
        left_arr.append(arr_temp[0])
        right_arr.insert(0, arr_temp[-1])
        arr = arr_temp
        length_arr -= 1
        mid_points.append(arr_temp)

    return [left_arr, [arr[0]], right_arr, mid_points]

def dnc_bezier(points, iteration, depth):
    if depth == 0:
        return {"points": [], "mid_points_history": []}

    left_side, mid, right_side, current_mid_points = iterate_find(points)
    left_result = dnc_bezier(left_side, iteration, depth - 1)
    right_result = dnc_bezier(right_side, iteration, depth - 1)

    return {
        "points": left_result["points"] + mid + right_result["points"],
        "mid_points_history": left_result["mid_points_history"] + [{"iteration": iteration - depth + 1, "mid_points": current_mid_points}] + right_result["mid_points_history"]
    }
